810-well-known-custom-extensions.md

```diff
@@ -0,0 +1,56 @@
+---
+title: Well known custom extensions
+weight: 810
+---
+
+# Well known custom actions
+
+This section is non-normative, but is here to propose a common set of custom extensions a bundle MAY define, and that CNAB tools MAY implement.
+
+A bundle indicates that it uses a custom extension by including them in its custom section (as defined in [Custom Extensions](101-bundle-json.md#custom-extensions)).
+
+## Dependencies
+
+See the [Dependencies Specification](500-CNAB-dependencies.md).
+
+## Parameter Sources
+
+A custom bundle extension, `io.cnab.parameterSources`, MAY be defined that provides additional guidance for how a runtime MAY determine the default value of a [parameter][parameter]. The parameter source extension does not cover how the runtime should set the value of a parameter because each implemenation may have different information to take into account when determining the final value of a parameter.
+
+In the example below, the [output][output] `tfstate` is initially generated during the install action, and the required parameter `tfstate` is used by the upgrade and uninstall actions. The parameter source specifies that the `tfstate` output can be used to set the `tfstate` parameter. This enables a bundle to pass data, in this case state, between actions.
+
+```json
+{
+  "custom": {
+    "io.cnab.parameterSources": {
+      "tfstate": {
+        "output": {
```

Comment:
Yes, I think adding more parameter sources over time is something we should be ready to support. If we ever decide to allow multiple sources to be specified for the same parameter, it might be a good idea to go ahead and nest the source configurations themselves so that we have somewhere to put additional configuration like relative priority without this being a breaking change.

For example, this is what it might look like to define a parameter that can be sourced from an output, but also supports fallback to a default value referenced as OCI via [ORAS](https://github.com/deislabs/oras).

```json
{
  "custom": {
    "io.cnab.parameter-sources": {
      "myparam": {
        "priority": ["output", "oras"],
        "sources": {
          "oras": {
            "image": "myparam-default"
          },
          "output": {
            "name": "myoutput"
          }
        }
      }
    }
  },
  "images": {
    "myparam-default": {
      "description": "fallback value for myparam",
      "image": "registry.example.com/hello-artifact:v1",
      "contentDigest": "sha256:asdf..."
    }
  }
}
```